### ISSUE TITLE
Catch-all pattern match

### ISSUE THREAD

**Issue by identicalsnowflake** [agda/agda#2124]
This isn't a bug per se, but it's an annoying issue which has come up repeatedly for me when using Agda. Consider the following, where everything past line 2 is basically pure mechanical repetition:

```
+-associative-valid : ∀ p → reifyReduction PolynomialReductions.+-associative p
+-associative-valid (Plus (Plus p p₁) p₂) = SemigroupLaws.associative +-semigroup
+-associative-valid (Plus (Const x) p₁) = refl
+-associative-valid (Plus (Var x) p₁) = refl
+-associative-valid (Plus (Times p p₁) p₂) = refl
+-associative-valid (Plus (Exp p x) p₁) = refl
+-associative-valid (Const x) = refl
+-associative-valid (Var x) = refl
+-associative-valid (Times p p₁) = refl
+-associative-valid (Exp p x) = refl
```

This is enormously verbose -- literally 90% of those patterns are simply syntactic noise. I want to write this:

```
+-associative-valid : ∀ p → reifyReduction PolynomialReductions.+-associative p
+-associative-valid (Plus (Plus p p₁) p₂) = SemigroupLaws.associative +-semigroup
+-associative-valid _ = refl
```

But Agda doesn't accept this, even though the path from the latter to the former is a purely mechanical process of me just doing "case split, write refl everywhere."

Another notoriously problematic case of this is writing decidable equality instances for any GADT with several constructors. DecEq requires N^2 lines of code (where N is the number of constructors), even though only N of them actually do anything!

Can we make Agda accept the latter, or something similar to it?

**Comment by ghost** [comment]
maybe agda should have something similar to coq's tactic that act on every branch.
something like:
induction e; refl.
in coq

**Comment by jespercockx** [comment]
One solution would be to expand catch-all patterns _before_ checking the right-hand side. However, this can cause seemingly ill-typed patterns to be accepted by Agda. It's also not clear what the type of a hole should be in this situation, e.g. what would be the type of `?` in 

```
+-associative-valid : ∀ p → reifyReduction PolynomialReductions.+-associative p
+-associative-valid (Plus (Plus p p₁) p₂) = SemigroupLaws.associative +-semigroup
+-associative-valid _ = ?
```

?

The other solution would be to leverage Agda's reflection mechanism to generate the boilerplate clauses automatically. However this still takes some effort to decide for which constructors to generate cases. You could take a look at DerivingEq in Ulf's prelude (https://github.com/UlfNorell/agda-prelude/blob/master/src/Tactic/Deriving/Eq.agda).

For this example it would be useful if we could write some clauses by hand and generate the rest by reflection. @UlfNorell, maybe this is something we could add?

**Comment by UlfNorell** [comment]
It think it's possible already. Here's an idea off the top of my head:

``` agda
-- To get f : A → B, you define this
f-template : ⊥ → A → B
f-template _ p₁ = e₁
f-template _ p₂ = e₂
f-template () _

-- Now I'm pretty sure we can define a function
fillInMissingCases : Name → Name → Tactic → TC ⊤
fillInMissingCases f template tac = ...

unquoteDecl f = fillInMissingCases f (quote f-template) reflTac
```

**Comment by andreasabel** [comment]
@jespercockx: Well, the context of the meta would exclude the variables coming from the automatic split.  In this case, the meta would be closed (not dependent on any variable).
The type of the meta should be `reifyReduction PolynomialReductions.+-associative p`, we have to do something about the `p` however.  Contextual TT to the rescue!  The substitution (specialization of p) stays outside the meta (an is invisible to the user).

Concretely, we can keep the shared rhs in a local definition which is instantiated in each clause generated by the split with appropriate arguments (representing the substitution).

**Comment by nad** [comment]
> DecEq requires N^2 lines of code (where N is the number of constructors), even though only N of them actually do anything!

No, you can (at least in many cases) get away with fewer cases. For instance, if you have a data type called `ASCII` with 128 nullary constructors, then you can first prove that `ASCII` is isomorphic to `Fin 128`, and then use decidable equality for `Fin 128`.

**Comment by EdNutting** [comment]
Has a reasonable way to solve this been found? If so, I'd love to know - I'm trying to engineer a system and having the ability to write "xyz _  = thing for all cases" would be extremely useful.

**Comment by jespercockx** [comment]
In theory --- yes. The elaboration algorithm described in our (Andreas and mine) upcoming [ICFP paper](https://jesper.sikanda.be/files/elaborating-dependent-copattern-matching.pdf) does case splitting *before* checking the right-hand side (or rather, interleaved with), so you get the effect you want. It's just a question of finding the time and mental stamina to do all the refactoring required to make it actually work in Agda.

**Comment by nad** [comment]
How do you handle open goals on the right-hand side?

Consider the following code:
```agda
data Three : Set where
  zero one two : Three

F : Three → Set
F zero = ⊥
F one  = ⊤
F two  = Bool

G : (x : Three) → F x → Set
G zero = ?
G _    = ?
```
What is the final goal type?

**Comment by gallais** [comment]
> Has a reasonable way to solve this been found?

The principled solution is to use a view. The explicit representation
of catchall as a constructor of the view means that the function will
reduce when it is told "and now this is the default case".

Cf. this example of a decidable equality proof which is linear in the
number of constructors for the datatype:

https://github.com/gallais/potpourri/blob/master/agda/poc/LinearDec.agda

Or these optimising functions for arithmetic expressions taken from
Software Foundations and the proofs that they are semantics preserving:
https://github.com/gallais/potpourri/blob/master/agda/sf/Imp.agda

Of course this does not work when the proof for the catchall case
needs to explicitly know that the expression did not match any of
the cases that came before it. But as you can see we can already do
quite a lot with this setup.
